hypershift-operator/controllers/scheduler/placeholders.go

```diff
@@ -0,0 +1,425 @@
+			wantedDeployments = sizeClass.Management.Placeholders
+		}
+	}
+
+	parsedIndex, err := strconv.Atoi(placeholderIndex)
+	if err != nil {
+		logger.Error(err, "deployment has invalid placeholder index value", "value", placeholderIndex)
+		return false, nil, nil
+	}
+
+	// index is zero-based, user intent is 1-based
+	if parsedIndex > wantedDeployments-1 {
+		// we have more deployments than we need, delete this one
+		return true, nil, nil
+	}
+
+	nodes, err := r.listNodes(ctx, client.HasLabels{HostedClusterNameLabel, OSDFleetManagerPairedNodesLabel})
+	if err != nil {
+		return false, nil, err
+	}
+
+	pairedNodesAssignedToHostedClusters := sets.Set[string]{}
+	for _, node := range nodes.Items {
+		pairedNodesAssignedToHostedClusters.Insert(node.ObjectMeta.Labels[OSDFleetManagerPairedNodesLabel])
+	}
+
+	existingPairedNodes := sets.Set[string]{}
+	for _, term := range deployment.Spec.Template.Spec.Affinity.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution.NodeSelectorTerms {
+		for _, expression := range term.MatchExpressions {
+			if expression.Key == OSDFleetManagerPairedNodesLabel && expression.Operator == corev1.NodeSelectorOpNotIn {
+				existingPairedNodes.Insert(expression.Values...)
+			}
+		}
+	}
+
+	if !pairedNodesAssignedToHostedClusters.Equal(existingPairedNodes) {
+		// the set of paired nodes on which hosted clusters have been scheduled has changed, we need to update the
+		// node affinity configuration on the deployment
+		return false, newDeployment(placeholderNamespace, placeholderSize, parsedIndex, pairedNodesAssignedToHostedClusters.UnsortedList()), nil
+	}
+	return false, nil, nil
+}
+
+func newDeployment(namespace, sizeClass string, placeholderIndex int, pairedNodes []string) *appsv1applyconfigurations.DeploymentApplyConfiguration {
+	var nodeAffinity *corev1applyconfigurations.NodeAffinityApplyConfiguration
+	if len(pairedNodes) > 0 {
+		// we can't add this unless there's something for the NotIn to match
+		nodeAffinity = corev1applyconfigurations.NodeAffinity().WithRequiredDuringSchedulingIgnoredDuringExecution(
+			corev1applyconfigurations.NodeSelector().WithNodeSelectorTerms(
+				// placeholder pods may not land on any nodes where other hosted clusters may be scheduled, even if
+				// the hosted control plane is currently not currently there - for instance, if hosted cluster is
+				// a 'large' size, we don't want placeholders landing on the 'small' nodes that the cluster would use
+				// if it were to scale down, since keeping those warm does not help us start new clusters more quickly
+				corev1applyconfigurations.NodeSelectorTerm().WithMatchExpressions(
+					corev1applyconfigurations.NodeSelectorRequirement().
+						WithKey(OSDFleetManagerPairedNodesLabel).
+						WithOperator(corev1.NodeSelectorOpNotIn).
+						WithValues(pairedNodes...),
+				),
+			),
+		)
+	}
+	return appsv1applyconfigurations.Deployment(strconv.Itoa(placeholderIndex), namespace).WithLabels(map[string]string{
```

Comment:
should the name be prefixed by the sizeClass? 
not a real use case, but your api allows it... if we have more than one size class with placeholders, they'll be overwriting each other